### PR TITLE
MULE-7990: Create Serialization API

### PR DESCRIPTION
**Pull request by marianogonzalez**

### CODE REVIEW

File: core/src/main/java/org/mule/api/serialization/DefaultObjectSerializer.java [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+import org.mule.api.config.MuleConfiguration;
+
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.Target;
+
+import javax.inject.Qualifier;
+
+/**
+ * A {@link Qualifier} annotation for injecting
```
Comment (pablokraan) [inline core/src/main/java/org/mule/api/serialization/DefaultObjectSerializer.java]: Adjust javadoc lines width

File: core/src/main/java/org/mule/api/serialization/ObjectSerializer.java [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+ * <p/>
+ * Implementations are required to be thread-safe
+ *
+ * @since 3.7.0
+ */
+public interface ObjectSerializer
+{
+
+    /**
+     * Serealizes the given object into a an array of {@link byte}s
```
Comment (pablokraan) [inline core/src/main/java/org/mule/api/serialization/ObjectSerializer.java]: Serealizes -> Serializes

File: core/src/main/java/org/mule/api/serialization/ObjectSerializer.java [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+     *
+     * @param object the object to be serialized. Might be <code>null</code>
+     * @return an array of {@link byte}
+     * @throws SerializationException in case of unexpected exception
+     */
+    byte[] serialize(Object object) throws SerializationException;
+
+    void serialize(Object object, OutputStream out) throws SerializationException;
```
Comment (pablokraan) [inline core/src/main/java/org/mule/api/serialization/ObjectSerializer.java]: Add javadoc

File: core/src/main/java/org/mule/api/serialization/ObjectSerializer.java [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+
+    /**
+     * Deserializes the given bytes. Unexpected behavior can result of deserializing
+     * a byte[] that was generated with another implementation.
```
Comment (pablokraan) [inline core/src/main/java/org/mule/api/serialization/ObjectSerializer.java]: I would put  "Unexpected behavior can result of deserializing a byte[] that was generated with another implementation"
in a different paragraph and clarify "with another {@link ObjectSerializer} implementation

File: core/src/main/java/org/mule/api/serialization/ObjectSerializer.java [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+     * <p/>
+     * If the obtained object implements {@link DeserializationPostInitialisable}
+     * then this serializer will be responsible for properly initializing
+     * the object before returning it.
+     * <p/>
+     * Implementation will choose the {@link java.lang.ClassLoader}
```
Comment (pablokraan) [inline core/src/main/java/org/mule/api/serialization/ObjectSerializer.java]: Why the implementation can choose the classloader to use? Should not be provided in the method invocation?

File: core/src/main/java/org/mule/api/serialization/ObjectSerializer.java [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+     * @param bytes an array of byte that an original object was serialized into
+     * @return the deserialized object
+     * @throws IllegalArgumentException if {@code bytes} is {@code null}
+     * @throws SerializationException   in case of unexpected exception
+     */
+    <T> T deserialize(byte[] bytes) throws SerializationException;
+
+    /**
+     * Deserializes the given bytes. Unexpected behavior can result of deserializing
+     * a byte[] that was generated with another implementation.
```
Comment (pablokraan) [inline core/src/main/java/org/mule/api/serialization/ObjectSerializer.java]: Maybe this text "Unexpected behavior can result of deserializing a byte[] that was generated with another implementation." could be in the class javadoc instead of repeating it in the different methods

File: core/src/main/java/org/mule/api/serialization/ObjectSerializerAware.java [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+ * Unlike similar interfaces like {@link MuleContextAware}, the platform
+ * will not automatically inject the default {@link ObjectSerializer} into
+ * instances of this interface. Recommended ways of using this interface
+ * is by either annotating the method with {@link Inject} or manually invoking it
+ * if the instance is not part of the registry and you don't
+ * want to use {@link MuleContext#getInjector()}
+ *
+ * @since 3.7.0
+ */
+public interface ObjectSerializerAware
```
Comment (pablokraan) [inline core/src/main/java/org/mule/api/serialization/ObjectSerializerAware.java]: Do we need another XXAware interface now that we can reliably inject objects?

File: core/src/test/java/org/mule/processor/IdempotentRedeliveryPolicyTestCase.java [hunk 8]
```diff
@@ -98,6 +100,9 @@ public ObjectStore answer(InvocationOnMock invocation) throws Throwable
             }
         });
         when(event.getMessage()).thenReturn(message);
+
+        IdempotentRedeliveryPolicyTestCase.serializer = SerializationTestUtils.getJavaSerializerWithMockContext();
```
Comment (pablokraan) [inline core/src/test/java/org/mule/processor/IdempotentRedeliveryPolicyTestCase.java]: This test already has a mulecontext, should not use that instance to configure the serializer?

File: core/src/test/java/org/mule/serialization/AbstractObjectSerializerContractTestCase.java [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+
+        MuleEvent event = getTestEvent(dateTime);
+        byte[] bytes = serializer.serialize(event.getMessage());
+
+        MuleMessage message = serializer.deserialize(bytes);
+        DateTime deserealized = (DateTime) message.getPayload();
+
+        assertEquals(calendar, deserealized.toCalendar());
+
+        // test that the locale matches
```
Comment (pablokraan) [inline core/src/test/java/org/mule/serialization/AbstractObjectSerializerContractTestCase.java]: remove redundant comment

File: core/src/test/java/org/mule/serialization/JavaObjectSerializerTestCase.java [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.serialization;
+
+import org.mule.api.serialization.SerializationException;
+
+import org.junit.Test;
+
+public class JavaObjectSerializerTestCase extends AbstractObjectSerializerContractTestCase
```
Comment (pablokraan) [inline core/src/test/java/org/mule/serialization/JavaObjectSerializerTestCase.java]: Is this a smallTest?

File: core/src/test/java/org/mule/tck/SerializationTestUtils.java [hunk 11]
```diff
@@ -45,4 +50,26 @@
...
+
+    public static MuleContext addJavaSerializerToMockMuleContext(MuleContext mockMuleContext)
+    {
+        when(mockMuleContext.getObjectSerializer()).thenReturn(getJavaSerializer(mockMuleContext));
+        return mockMuleContext;
+    }
+
+    private static ObjectSerializer getJavaSerializer(MuleContext muleContext)
+    {
+        JavaObjectSerializer serializer = new JavaObjectSerializer();
```
Comment (pablokraan) [inline core/src/test/java/org/mule/tck/SerializationTestUtils.java]: Check if muleContext serializer was already initialised

File: core/src/test/java/org/mule/util/queue/DualRandomAccessFileQueueStoreDelegateTestCase.java [hunk 12]
```diff
@@ -34,6 +35,16 @@
     @Rule
     public TemporaryFolder workingDirectory = new TemporaryFolder();
 
+    private MuleContext mockMuleContext;
+
+    @Before
+    public void before()
+    {
+        mockMuleContext = Mockito.mock(MuleContext.class);
```
Comment (pablokraan) [inline core/src/test/java/org/mule/util/queue/DualRandomAccessFileQueueStoreDelegateTestCase.java]: Use static import

File: modules/spring-config/src/main/java/org/mule/config/spring/DefaultObjectSerializerFactoryBean.java [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+ * {@link DefaultObjectSerializerWrapper} so that the instance returned by this
+ * factory is one annotated with {@link DefaultObjectSerializer}. Ideally, that
+ * shouldn't be necessary since this class also has that annotation and that should
+ * be enough. However, due to <a href="https://jira.spring.io/browse/SPR-12914">
+ * Spring issue SPR-12914</a> we have to annotate both classes, since Spring
+ * will only consider the annotation in this class when injecting in runtime,
+ * while the annotation on {@link DefaultObjectSerializerWrapper} will only be
+ * considered at startup time.
+ * <p/>
+ * When Spring fixes
```
Comment (pablokraan) [inline modules/spring-config/src/main/java/org/mule/config/spring/DefaultObjectSerializerFactoryBean.java]: Fix javadoc line width

File: tests/integration/src/test/java/org/mule/config/spring/InjectDefaultObjectSerializerTestCase.java [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+
+    @Test
+    public void injectObjectSerializer() throws Exception
+    {
+        TestObjectSerializerAware objectSerializerAware = muleContext.getInjector().inject(new TestObjectSerializerAware());
+        assertThat(muleContext.getObjectSerializer(), is(sameInstance(objectSerializerAware.getObjectSerializer())));
+        assertThat(objectSerializerAware.getObjectSerializer(), is(sameInstance(muleContext.getRegistry().get(MuleProperties.OBJECT_SERIALIZER))));
+    }
+
+    public static class TestObjectSerializerAware
```
Comment (pablokraan) [inline tests/integration/src/test/java/org/mule/config/spring/InjectDefaultObjectSerializerTestCase.java]: Why TestObjectSerializerWare if it does not implements ObjectSerializerAware?

File: transports/tcp/src/main/java/org/mule/transport/tcp/TcpConnector.java [hunk 15]
```diff
@@ -154,6 +154,18 @@ private boolean newValue(int parameter, int socketValue)
     @Override
     protected void doInitialise() throws InitialisationException
     {
+        if (tcpProtocol != null)
+        {
+            try
+            {
+                muleContext.getInjector().inject(getTcpProtocol());
```
Comment (pablokraan) [inline transports/tcp/src/main/java/org/mule/transport/tcp/TcpConnector.java]: Why is this needed?

File: core/src/main/java/org/mule/api/serialization/DefaultObjectSerializer.java (hunk 1 above)
Comment (marianogonzalez) [inline core/src/main/java/org/mule/api/serialization/DefaultObjectSerializer.java]: kinda fixed

File: core/src/main/java/org/mule/api/serialization/ObjectSerializer.java (hunk 2 above)
Comment (marianogonzalez) [inline core/src/main/java/org/mule/api/serialization/ObjectSerializer.java]: fixed

File: core/src/main/java/org/mule/api/serialization/ObjectSerializer.java (hunk 3 above)
Comment (marianogonzalez) [inline core/src/main/java/org/mule/api/serialization/ObjectSerializer.java]: fixed

File: core/src/main/java/org/mule/api/serialization/ObjectSerializer.java (hunk 4 above)
Comment (marianogonzalez) [inline core/src/main/java/org/mule/api/serialization/ObjectSerializer.java]: fixed and moved to the class javadoc to avoid repetition

File: core/src/main/java/org/mule/api/serialization/ObjectSerializer.java (hunk 5 above)
Comment (marianogonzalez) [inline core/src/main/java/org/mule/api/serialization/ObjectSerializer.java]: there's an overload version in which you can specify the classloader.

File: core/src/main/java/org/mule/api/serialization/ObjectSerializer.java (hunk 6 above)
Comment (marianogonzalez) [inline core/src/main/java/org/mule/api/serialization/ObjectSerializer.java]: moved up

File: core/src/main/java/org/mule/api/serialization/ObjectSerializerAware.java (hunk 7 above)
Comment (marianogonzalez) [inline core/src/main/java/org/mule/api/serialization/ObjectSerializerAware.java]: This is for the cases in which after injection you want to propagate the serializer further into nested objects which are not exposed. See SafeProtocol for an example.

File: core/src/test/java/org/mule/processor/IdempotentRedeliveryPolicyTestCase.java (hunk 8 above)
Comment (marianogonzalez) [inline core/src/test/java/org/mule/processor/IdempotentRedeliveryPolicyTestCase.java]: doing that interfeers with another test in this class and doing this as it is doesn't hurt since everything is a big mock

File: core/src/test/java/org/mule/serialization/AbstractObjectSerializerContractTestCase.java (hunk 9 above)
Comment (marianogonzalez) [inline core/src/test/java/org/mule/serialization/AbstractObjectSerializerContractTestCase.java]: done

File: core/src/test/java/org/mule/serialization/JavaObjectSerializerTestCase.java (hunk 10 above)
Comment (marianogonzalez) [inline core/src/test/java/org/mule/serialization/JavaObjectSerializerTestCase.java]: no because the base class extends AbstractMuleContextTestCase

File: core/src/test/java/org/mule/tck/SerializationTestUtils.java (hunk 11 above)
Comment (marianogonzalez) [inline core/src/test/java/org/mule/tck/SerializationTestUtils.java]: this is just a utility class for tests.. I don't think we care about that here

File: core/src/test/java/org/mule/util/queue/DualRandomAccessFileQueueStoreDelegateTestCase.java (hunk 12 above)
Comment (marianogonzalez) [inline core/src/test/java/org/mule/util/queue/DualRandomAccessFileQueueStoreDelegateTestCase.java]: fixed

File: modules/spring-config/src/main/java/org/mule/config/spring/DefaultObjectSerializerFactoryBean.java (hunk 13 above)
Comment (marianogonzalez) [inline modules/spring-config/src/main/java/org/mule/config/spring/DefaultObjectSerializerFactoryBean.java]: kinda fixed

File: tests/integration/src/test/java/org/mule/config/spring/InjectDefaultObjectSerializerTestCase.java (hunk 14 above)
Comment (marianogonzalez) [inline tests/integration/src/test/java/org/mule/config/spring/InjectDefaultObjectSerializerTestCase.java]: doesn't need to, its using DI

File: transports/tcp/src/main/java/org/mule/transport/tcp/TcpConnector.java (hunk 15 above)
Comment (marianogonzalez) [inline transports/tcp/src/main/java/org/mule/transport/tcp/TcpConnector.java]: Because the protocol does not go into the registry but it does need to have the serializer injected anyway. So it's either this or checking for ObjectSerializerAware
